server/src/main/java/org/opensearch/plugins/IdentityPlugin.java

```diff
@@ -29,4 +30,9 @@ public interface IdentityPlugin {
      * @return Should never return null.
      */
     public TokenManager getTokenManager();
+
+    /**
+     * Attempt to authenticate the provided subject using the provided authToken
+     */
+    public void authenticate(Subject subject, AuthToken authToken);
```

Comment:
We shouldn't elevate this method from the subject to the plugin - let's keep it on the subject. 

---

I'm pretty sure we are missing documentation/flow chart/diagram that expresses the lifecycle of subject inside OpenSearch, I'll attempt some of that in a comment now and I'll follow up with an issue.

'Identity' is should always be determinable within OpenSearch, and it's associated with the context of an operation, e.g. during a request, first only the remote IP address is known, then maybe referral ip or x-forwarded source, then after checking the authentication header.  It's always the same subject, but the internal fidelity is increasing.

For this reason, the subject is mutating as data is collected, so authenticating with a token is adding more identity information - principals to the subject - not the identity plugin.

An identity plugin knows how to map different kinds of authentication tokens into principals, it also knows how to determine what the current subject is the context.  The security plugin uses the ThreadContext with several keys as the storage medium for these principals.

Reply:
So I raised this into the `IdentityPlugin` interface because I think the way this `authenticate` method was intended to be used was to set the active subject.

As I interpret the code, this authenticate was intended to be used to update the context for the cluster when a new operation occurred. I did not think it meant the traditional "attempt to login" type of authentication since that would be handled before the Subject was ever formed. You wouldn't start acting as subject 1 before you have logged in subject 1. So I figured this was intended to be used as a hook to update the context under which the cluster was operating. 

I then rose the level this method was used at because it did not make sense to me why we would have the subjects authenticate--update the context which was being used--when that is a tracking variable at the IdentityPlugin level (see `getSubject`). 

I agree we need some clearer details on this. Personally, I would like to see some of this functionality dropped for the meantime since it forces changes in the Security Plugin which are convoluted. We don't have any concept of a tracked Subject inside the Security Plugin right now and by adding this idea into the Identity Plugin we block OBO and Service Accounts which use the IdentityPlugin TokenManager. 

I am up to confirming the implementation of everything to Step 8 in this [diagram](https://github.com/opensearch-project/security/blob/main/ARCHITECTURE.md#extension-service-account-authorization), but am struggling to find a way to handle the subject use in the Security Plugin where we do not have any generic tracking for that logic.

Reply:
Removed the escalation